File: leetcode-js/123-best-time-to-buy-and-sell-stock-iii.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function (prices) {
    let holding1 = -Infinity;
    let holding2 = -Infinity;
    let release1 = 0;
    let release2 = 0;

    for (const p of prices) {
        // 1. Buy in at lowest price
        holding2 = Math.max(holding2, release1 - p);

        // 2. Sold out at highest price
        release2 = Math.max(release2, holding2 + p);

        holding1 = Math.max(holding1, -p);
        release1 = Math.max(release1, holding1 + p);
    }

    return release2;
};

console.log(maxProfit([1,2]));

File: leetcode-java/123-best-time-to-buy-and-sell-stock-iii.java
public class Solution {
    public int maxProfit(int[] prices) {
        int currentMax1 = 0;
        int max1 = 0;
        int currentMax2 = 0;
        int max2 = 0;

        for (int i = 1; i < prices.length; i++) {
            int diff = prices[i] - prices[i - 1];
            currentMax2 = Math.max(currentMax2 + diff, max1 + diff);
            max2 = Math.max(max2, currentMax2);
            currentMax1 = Math.max(currentMax1 + diff, diff);
            max1 = Math.max(max1, currentMax1);
        }

        return max2;
    }
}
